ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java

```diff
@@ -77,19 +75,19 @@
   private static final int SERVICE_TIMEOUT_SEC = 2;
   private static final int CHECK_HEARTBEAT_DELAY_MS = 1000;
   private static final int SEND_HEARTBEAT_DELAY_MS = 100;
+  private static final int DISCOVER_CLUSTER_DELAY_MS = 50;
   private static final Logger LOG = LoggerFactory.getLogger(HeartbeatAgent.class);
 
   private final KsqlEngine engine;
   private final ServiceContext serviceContext;
   private final HeartbeatConfig config;
   private final List<HostStatusListener> hostStatusListeners;
-  private final ConcurrentHashMap<String, TreeMap<Long, HeartbeatInfo>> receivedHeartbeats;
-  private final ConcurrentHashMap<String, HostStatusEntity> hostsStatus;
+  private final ConcurrentHashMap<KsqlHost, TreeMap<Long, HeartbeatInfo>> receivedHeartbeats;
+  private final AtomicReference<ImmutableMap<KsqlHost, HostStatus>> hostsStatus;
```

Comment:
to truly enforce immutability, it might (cc @big-andy-coates) make sense to make a class specifically wrapping `ImmutableMap<KsqlHost, HostStatus>` and mark it `@Immutable`. That way the `ImmutableTester` will make sure it is recursively immutable.

Also, wrapping this class could help us with handling the localhost in a cleaner way - you could initialize it passing in the local `KsqlHost` and handle it as an edge case in the `get()` method. Same goes for some other complicated logic that could be pushed into this class.

This can be done as a follow-up to unblock your other PRs.

Reply:
After more discussions and thought, the map should not be immutable. I am changing it back to be a concurrent map of immutable objects

Reply:
Can you explain your reasoning?

Reply:
We need finer-grained control over the map (as per your next comment). Moreover, ensuring immutability per key, value is enough